src/native/eventpipe/ep-session.c

```diff
@@ -183,25 +205,35 @@ static
 void
 session_disable_streaming_thread (EventPipeSession *session)
 {
-	EP_ASSERT (session->session_type == EP_SESSION_TYPE_IPCSTREAM || session->session_type == EP_SESSION_TYPE_FILESTREAM);
+	EP_ASSERT (ep_session_type_uses_streaming_thread (session->session_type));
 	EP_ASSERT (ep_session_get_streaming_enabled (session));
 
 	EP_ASSERT (!ep_rt_process_detach ());
-	EP_ASSERT (session->buffer_manager != NULL);
+	EP_ASSERT (!ep_session_type_uses_buffer_manager (session->session_type) || session->buffer_manager != NULL);
 
 	// The streaming thread will watch this value and exit
 	// when profiling is disabled.
 	ep_session_set_streaming_enabled (session, false);
 
 	// Thread could be waiting on the event that there is new data to read.
-	ep_rt_wait_event_set (ep_buffer_manager_get_rt_wait_event_ref (session->buffer_manager));
+	if (ep_session_type_uses_buffer_manager (session->session_type))
+		ep_rt_wait_event_set (ep_buffer_manager_get_rt_wait_event_ref (session->buffer_manager));
 
 	// Wait for the streaming thread to clean itself up.
 	ep_rt_wait_event_handle_t *rt_thread_shutdown_event = &session->rt_thread_shutdown_event;
 	ep_rt_wait_event_wait (rt_thread_shutdown_event, EP_INFINITE_WAIT, false /* bAlertable */);
 	ep_rt_wait_event_free (rt_thread_shutdown_event);
 }
 
+static
+bool
+session_is_stream_connection_closed (IpcStream *stream)
+{
+	EP_ASSERT (stream != NULL);
+	DiagnosticsIpcPollEvents poll_event = ep_ipc_stream_poll_vcall (stream, DS_IPC_TIMEOUT_INFINITE);
+	return poll_event == DS_IPC_POLL_EVENTS_HANGUP;
```

Comment:
Should we handle errors as close as well?

Reply:
I wasn't sure previously if there was a chance there could be an error from a poll operation even though the connection didn't close where a following poll retry would succeed. I think we can default to just treating both Error and hangup as connection closed.